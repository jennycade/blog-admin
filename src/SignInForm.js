import { useState } from 'react';

function SignInForm(props) {
  // props
  const {
    handleSignInSubmit,
  } = props;

  // state
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const [error, setError] = useState('');

  // functions
  const onUsernameChange = (event) => {
    setUsername(event.target.value);
  }
  const onPasswordChange = (event) => {
    setPassword(event.target.value);
  }
  const handleSubmit = async (event) => {
    event.preventDefault();

    // form validation?
    if (username === '') {
      setError('Username required');
    } else if (password === '') {
      setError('Password required');
    } else {
      await handleSignInSubmit(username, password);
    }
  }

  return (
    <form onSubmit={handleSubmit}>
      { error !== '' && (
        <p role='alert' className='alert alert-danger'>
          {error}
        </p>
      )}

      <div className='mb-3'>
        <label
          htmlFor="usernameInput"
          className='form-label'
        >Username</label>
        <input
          type='text'
          className='form-control'
          id='usernameInput'
          value={username}
          onChange={onUsernameChange}
        />
      </div>

      <div className='mb-3'>
        <label
          htmlFor="passwordInput"
          className='form-label'
        >Password</label>
        <input
          type='password'
          className='form-control'
          id='passwordInput'
          value={password}
          onChange={onPasswordChange}
        />
      </div>
      <div className='mb-3'>
        <button className='btn btn-primary'>Sign in</button>
      </div>
    </form>
  );
}

export default SignInForm;