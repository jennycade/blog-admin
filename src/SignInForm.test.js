import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import SignInForm from './SignInForm';

describe('SignInForm component', () => {
  it('matches previous snapshot', () => {
    const { container } = render(<SignInForm />);
    expect(container).toMatchSnapshot();
  });

  it('shows sign in form on load', () => {
    render(<SignInForm />);
    expect(screen.getByRole('textbox')).toBeInTheDocument();
  });

  it('updates username when user types', async () => {
    const user = userEvent.setup();
    render(<SignInForm />);
    
    const usernameField = screen.getByLabelText('Username');
    await user.type(usernameField, 'user');

    expect(usernameField.value).toBe('user');
  });

  it('updates password when user types', async () => {
    const user = userEvent.setup();
    render(<SignInForm />);
    
    const passwordField = screen.getByLabelText('Password');
    await user.type(passwordField, 'abc123');

    expect(passwordField.value).toBe('abc123');
  });

  it('calls handleSignInSubmit() prop with username and password when sign in button is clicked', async () => {
    const user = userEvent.setup();
    // mock submit
    const handleSignInSubmit = jest.fn();
    render(<SignInForm handleSignInSubmit={handleSignInSubmit} />);

    // UI
    const usernameField = screen.getByLabelText('Username');
    const passwordField = screen.getByLabelText('Password');
    const submitButton = screen.getByRole('button', { name: 'Sign in' });

    // fill out form
    await user.type(usernameField, 'testuser');
    await user.type(passwordField, 'abc123');

    // submit (click)
    await user.click(submitButton);

    // assertion
    expect(handleSignInSubmit.mock.calls.length).toBe(1);
    expect(handleSignInSubmit.mock.calls[0][0]).toBe('testuser');
    expect(handleSignInSubmit.mock.calls[0][1]).toBe('abc123');
  });

  it('calls handleSignInSubmit() prop with username and password with enter key', async () => {
    const user = userEvent.setup();
    // mock submit
    const handleSignInSubmit = jest.fn();
    render(<SignInForm handleSignInSubmit={handleSignInSubmit} />);

    // UI
    const usernameField = screen.getByLabelText('Username');
    const passwordField = screen.getByLabelText('Password');

    // fill out form
    await user.type(usernameField, 'testuser');
    await user.type(passwordField, 'abc123');

    // submit (enter)
    await user.keyboard('{Enter}');

    // assertion
    expect(handleSignInSubmit.mock.calls.length).toBe(1);
    expect(handleSignInSubmit.mock.calls[0][0]).toBe('testuser');
    expect(handleSignInSubmit.mock.calls[0][1]).toBe('abc123');
  });

  // validation
  it('does not call handleSignInSubmit if username is missing', async () => {
    const user = userEvent.setup();
    // mock submit
    const handleSignInSubmit = jest.fn();
    render(<SignInForm handleSignInSubmit={handleSignInSubmit} />);

    // UI
    // const usernameField = screen.getByLabelText('Username');
    const passwordField = screen.getByLabelText('Password');

    // fill out form
    // await user.type(usernameField, 'testuser');
    await user.type(passwordField, 'abc123');

    // submit (enter)
    await user.keyboard('{Enter}');

    // assertion
    expect(handleSignInSubmit.mock.calls.length).toBe(0);
  });
  it('does not call handleSignInSubmit if password is missing', async () => {
    const user = userEvent.setup();
    // mock submit
    const handleSignInSubmit = jest.fn();
    render(<SignInForm handleSignInSubmit={handleSignInSubmit} />);

    // UI
    const usernameField = screen.getByLabelText('Username');
    // const passwordField = screen.getByLabelText('Password');

    // fill out form
    await user.type(usernameField, 'testuser');
    // await user.type(passwordField, 'abc123');

    // submit (enter)
    await user.keyboard('{Enter}');

    // assertion
    expect(handleSignInSubmit.mock.calls.length).toBe(0);
  });

  it('shows an error message with the term "username" if username is missing', async () => {
    const user = userEvent.setup();
    // mock submit
    const handleSignInSubmit = jest.fn();
    render(<SignInForm handleSignInSubmit={handleSignInSubmit} />);

    // UI
    // const usernameField = screen.getByLabelText('Username');
    const passwordField = screen.getByLabelText('Password');

    // fill out form
    // await user.type(usernameField, 'testuser');
    await user.type(passwordField, 'abc123');

    // submit (enter)
    await user.keyboard('{Enter}');

    // assertions
    const alert = screen.getByRole('alert');
    expect(alert).toBeInTheDocument();
    expect(alert.textContent).toMatch(/username/i);
  });

});