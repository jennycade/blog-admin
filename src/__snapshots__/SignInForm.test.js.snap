// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInForm component matches previous snapshot 1`] = `
<div>
  <form>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="usernameInput"
      >
        Username
      </label>
      <input
        class="form-control"
        id="usernameInput"
        type="text"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <label
        class="form-label"
        for="passwordInput"
      >
        Password
      </label>
      <input
        class="form-control"
        id="passwordInput"
        type="password"
        value=""
      />
    </div>
    <div
      class="mb-3"
    >
      <button
        class="btn btn-primary"
      >
        Sign in
      </button>
    </div>
  </form>
</div>
`;
